# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1233871449 18000
#      Thu Feb 05 17:04:09 2009 -0500
# Node ID f3579c1d6be98365af149dd00b966b3a3d94915d
# Parent  dd13355fc98a68cddca4bb82285e12b113b35271
Fdrawnow: don't strip dirsep; return after errors

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,10 +1,13 @@
 2009-02-05  John W. Eaton  <jwe@octave.org>
 
+	* graphics.cc (Fdrawnow): Return after errors.  Don't strip
+	trailing directory separator from name used in call to file_stat.
+
 	* DLD-FUNCTIONS/eigs.cc (Feigs): Avoid warnings about ambiguous
 	else.  Style fixes.  Return after errors.
 
 2009-02-05  Jaroslav Hajek  <highegg@gmail.com>
 
 	* oct-map.cc (Octave_map::index): Optimize.
 
 2009-02-05  Jaroslav Hajek  <highegg@gmail.com>
diff --git a/src/graphics.cc b/src/graphics.cc
--- a/src/graphics.cc
+++ b/src/graphics.cc
@@ -4979,17 +4979,20 @@ Undocumented internal function.\n\
 
 	  if (args.length () == 1)
 	    {
 	      caseless_str val (args(0).string_value ());
 
 	      if (! error_state && val.compare ("expose"))
 		do_events = false;
 	      else
-		error ("drawnow: invalid argument, expected `expose' as argument");
+		{
+		  error ("drawnow: invalid argument, expected `expose' as argument");
+		  return retval;
+		}
 	    }
 
 	  if (do_events)
 	    {
 	      gh_manager::unlock ();
 
 	      gh_manager::process_events ();
 
@@ -5008,21 +5011,27 @@ Undocumented internal function.\n\
 	      file = args(1).string_value ();
 
 	      if (! error_state)
 		{
 		  size_t pos = file.find_last_of (file_ops::dir_sep_chars ());
 
 		  if (pos != std::string::npos)
 		    {
-		      file_stat fs (file.substr (0, pos));
+		      std::string dirname = file.substr (0, pos+1);
+
+		      file_stat fs (dirname);
 
 		      if (! (fs && fs.is_dir ()))
-			error ("drawnow: nonexistent directory `%s'",
-			       file.substr (0, pos).c_str ());
+			{
+			  error ("drawnow: nonexistent directory `%s'",
+				 dirname.c_str ());
+
+			  return retval;
+			}
 		    }
 
 		  mono = (args.length () >= 3 ? args(2).bool_value () : false);
 
 		  if (! error_state)
 		    {
 		      debug_file = (args.length () > 3 ? args(3).string_value ()
 				    : "");
